src/webmon_app/reporting/report/view_util.py

```diff
@@ -249,45 +249,54 @@ def _get_rate(id_name):
         Returns rate for runs or errors
         :param id_name: 'run' or 'error'
         """
-        rate = cache.get("%s_%s_rate" % (instrument_id.name, id_name))
-        if rate is not None and last_cached_run is not None and last_run == last_cached_run:
+        # do not bother with the cache if it is invalid
+        if last_cached_run is not None and last_run == last_cached_run:
+            # this should either return the cache information or None
             return cache.get("%s_%s_rate" % (instrument_id.name, id_name))
-        return None
+        else:  # the cache is not valid
+            return None
 
     runs = _get_rate("run")
     errors = _get_rate("error")
 
-    # If we didn't find good rates in the cache, recalculate them
+    # If we didn't find useful information in the cache, recalculate them
     if runs is None or errors is None:
         # check for any run in the last n-hours
         time_oldest = timezone.now() - datetime.timedelta(hours=n_hours + 1)  # is +1 needed?
 
-        # only query further if there are runs for the instrument
-        have_runs = bool(DataRun.objects.filter(instrument_id=instrument_id, created_on__gte=time_oldest).count() > 0)
-        if have_runs:
-            runs = run_rate(instrument_id, n_hours=n_hours)
-        else:
-            runs = __generate_empty_rates(n_hours)
+        # only calculate if there isn't a cache for runs
+        if runs is None:
+            # only query further if there are runs for the instrument
+            have_runs = bool(
+                DataRun.objects.filter(instrument_id=instrument_id, created_on__gte=time_oldest).count() > 0
+            )
+            if have_runs:
+                runs = run_rate(instrument_id, n_hours=n_hours)
+            else:
+                runs = __generate_empty_rates(n_hours)
+
+                # cache the run rate
+                cache.set("%s_run_rate" % instrument_id.name, runs, settings.RUN_RATE_CACHE_TIMEOUT)
```

Comment:
should this be indented one level up so it's not under the `else` clause?

Reply:
Good eyes!